src/Microsoft.AspNetCore.Blazor.Build/Core/BootJsonWriter.cs

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.JSInterop;
+using Mono.Cecil;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace Microsoft.AspNetCore.Blazor.Build
+{
+    internal class BootJsonWriter
+    {
+        public static void WriteFile(
+            string assemblyPath,
+            string[] assemblyReferences,
+            string[] embeddedResourcesSources,
+            bool linkerEnabled,
+            string outputPath)
+        {
+            var embeddedContent = EmbeddedResourcesProcessor.ExtractEmbeddedResources(
+                embeddedResourcesSources, Path.GetDirectoryName(outputPath));
+            var bootJsonText = GetBootJsonContent(
+                Path.GetFileName(assemblyPath),
+                GetAssemblyEntryPoint(assemblyPath),
+                assemblyReferences,
+                embeddedContent,
+                linkerEnabled);
+            var normalizedOutputPath = Normalize(outputPath);
+            Console.WriteLine("Writing boot data to: " + normalizedOutputPath);
+            File.WriteAllText(normalizedOutputPath, bootJsonText);
+        }
+
+        public static string GetBootJsonContent(string assemblyFileName, string entryPoint, string[] assemblyReferences, IEnumerable<EmbeddedResourceInfo> embeddedContent, bool linkerEnabled)
+        {
+            var data = new BootJsonData(
+                assemblyFileName,
+                entryPoint,
+                assemblyReferences,
+                embeddedContent,
+                linkerEnabled);
+            return Json.Serialize(data);
+        }
+
+        private static string GetAssemblyEntryPoint(string assemblyPath)
+        {
+            using (var assemblyDefinition = AssemblyDefinition.ReadAssembly(assemblyPath))
```

Comment:
meta: do we need cecil for this? is the reason why we need to do this because otherwise we'd need to load the assembly?

Reply:
Yeah, loading assemblies in tooling, or at build time is just a fail. We shouldn't try. The `System.Reflection.Metadata` APIs are pretty low level and hard to use, hence Mono.Cecil still is the best option for many of these kinds of tasks.